src/phase2.hpp

```diff
@@ -52,14 +52,15 @@ Phase2Results RunPhase2(
     uint32_t const log_num_buckets,
     bool const show_progress)
 {
-    // An extra bit is used, since we may have more than 2^k entries in a table. (After pruning,
-    // each table will have 0.8*2^k or fewer entries).
+    // After pruning each table will have 0.865 * 2^k or fewer entries on
+    // average
```

Comment:
I observed empirically that the number of entries in tables 1-6 closely approximates the theoretical expected values. I also did the math to confirm that the probability of exceeding 2^k entries in those tables is negligible. Below is a calculation for table 6:

Assuming the table has `n = 2^k` entries after phase 1 and that `2^k` random pairs of entries match, the probability of an entry _not having_ a match is `((n-1)/n) ** (2*n) = (1 - 1/n) ** (2*n)`, which is approximated by `e^(-2) = 0.135` for a sufficiently large `n`. The probability of an entry _having_ a match (and thus remaining in table 6) is `p = 1 - 0.135 = 0.865`.

Now we can use the central limit theorem to prove that for a sufficiently large `n`, the number of matching entries `m` can be approximated by a normal distribution with `mean = n * p = 2^k * 0.865` and `stdev = sqrt(n * p * (1 - p)) = sqrt(2^k * 0.865 * (1 - 0.865)) = 2^(k/2) * 0.342`. The number of standard deviations required to reach `2^k` is given by `d = (2^k - mean) / stdev = 2^k * 0.135 / (2^(k/2) * 0.342) = 2^(k/2) * 0.396`. For the minimal `k = 18` we have `d = 2^9 * 0.395 = 202`, and it grows exponentially with larger `k`. It is clear that the probability of deviating from the mean by more than 200 sigma is so small that it can be considered impossible.

The number of entries in tables 1-5 after pruning is smaller than in table 6, and similar calculations also apply.